### ISSUE TITLE
currentDrawable issue

### ISSUE THREAD

**Issue by vjdaemp** [Harley-xk/MaLiang#78]
The Canvas' snapshot function simply renders the UIView to UIImage.  I prefer to convert the currentDrawable texture to UIImage directly. I am able to successfully use this technique to convert drawings to images. See code below:

`            
            self.canvas!.draw()
        
            // Convert the texture into an image
            let context = CIContext()
            let texture = self.canvas?.currentDrawable!.texture
            
            let cImg = CIImage(mtlTexture: texture!, options: nil)!
            let cgImg = context.createCGImage(cImg, from: cImg.extent)!
            let canvasCapture : UIImage? = UIImage(cgImage: cgImg, scale: 1.0, orientation: UIImage.Orientation.downMirrored)`

To use technique above, ensure framebufferOnly is set to false

`canvas!.framebufferOnly = false`

The issue is that when I tap on the screen to draw dots, the tap does not always get saved to texture, and do not appear in resulting UIImage until I draw a line.  I believe your code is rendering taps and lines differently , and taps are not getting saved to texture.

Can you replicate this? Do you have any ideas for a fix?

Thanks

**Comment by Harley-xk** [comment]
@vjdaemp There is a set of tool functions on MTLTexture for converting it to images, refer to the **MLTexture.swift** for details. Just call `canvas.currentDrawable.texture.toUIImage()` instead, it's the same as your code.
Call `canvas.setNeedsDisplay()` before taking a snapshot.
DO NOT set `framebufferOnly` to `false`, this will cause performance issue.

**Comment by vjdaemp** [comment]
@Harley-xk Thanks for the suggestion to use the simpler `canvas.currentDrawable.texture.toUIImage()` function instead of the code I was using.

If I do not set `framebufferOnly` to false, then I receive the following error:
`-[MTLDebugComputeCommandEncoder setTexture:atIndex:]:362: failed assertion `frameBufferOnly texture not supported for compute.`'

FYI, below are other properties that I have set for canvas:
```
        canvas!.isPaused = true
        canvas!.enableSetNeedsDisplay = false
```

I added the `canvas.setNeedsDisplay` before calling the function to convert texture to UIImage, but the same behavior is occurring. Is there a difference between `renderTap` and `pushPoint`. Are `renderTap` drawings not sent to texture right away?

**Comment by vjdaemp** [comment]
@Harley-xk UPDATE: I can replicate the issue on the Example app. If you add the following line to `viewDidLoad` in `ViewController.swift`:

`canvas.framebufferOnly = false`

and replace `snapshotAction` code with:

```
    func snapshotAction(_ sender: Any) {
        
        canvas.setNeedsDisplay()
        let canvasCapture = (canvas!.currentDrawable?.texture.toUIImage())!
        UIImageWriteToSavedPhotosAlbum(canvasCapture, nil, nil, nil)
    }
```

Draw a line on the canvas, then tap to draw a dot.  Invoke "Snapshot" action and observe the saved image does not include the dot.

**Comment by Harley-xk** [comment]
@vjdaemp Yes, I replicated your problem. But `renderTap` and `pushPoint` both calls the `renderLines` method on the bottom, `currentDrawable` property on `MTKView` may have some magic that causes this problem, I will find out it later.
There's another way to get a snapshot image from canvas' texture:  `SnapshotTarget` is designed to take snapshot on `ScrollableCanvas`, but is's also available on normal canvas.
You can take image from canvas by the code below:
```swift
let target = SnapshotTarget(canvas: canvas)
let image = target.getImage()
```
SnapshotTarget will make a new empty texture and copy the contents from the specified `canvas` object. By this way, there's no need to call `setNeedsDisplay` or to set `frameBufferOnly` to `false`.

**Comment by vjdaemp** [comment]
@Harley-xk Thanks for pointing me to `SnapshotTarget`. This method worked very well. As you mentioned, I was not required to set `framebufferOnly` to false.

If both renderTap and pushPoint use the same underlying method , then perhaps it is some issue with Apple Metal code/behavior. Do you think it might be a timing issue related to how the underlying views and textures get updated? What happens differently when a line is rendered vs a tap?
